### PR TITLE
Issue 140: Support lower case routes

### PR DESCRIPTION
**Pull request by kulmugdha**
@rynowak Please have a look.

### CODE REVIEW

**Comment by rynowak** [comment]
Please rebase on top of `dev`, I already added access to `RouteOptions` to the route collection

File: src/Microsoft.AspNet.Routing/RouteCollection.cs [hunk 1]
```diff
@@ -19,6 +20,11 @@ public class RouteCollection : IRouteCollection
         private readonly Dictionary<string, INamedRouter> _namedRoutes =
                                     new Dictionary<string, INamedRouter>(StringComparer.OrdinalIgnoreCase);
         private ILogger _logger;
+        private RouteOptions _routeOptions;
+        public RouteCollection(IOptions<RouteOptions> routeOptions)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouteCollection.cs]: spacing, let it breathe :relaxed:

File: src/Microsoft.AspNet.Routing/RouteCollection.cs [hunk 2]
```diff
@@ -135,7 +145,28 @@ public virtual string GetVirtualPath(VirtualPathContext context)
             return null;
         }
 
-        private void EnsureLogger(HttpContext context)
+        private string NormalizeVirtualPath(String url)
+        {
+            if (_routeOptions.LowercaseUrls)
+            {
+                string lowercaseUrl = url;
+                string queryString = "";
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouteCollection.cs]: var

File: src/Microsoft.AspNet.Routing/RouteCollection.cs [hunk 3]
```diff
@@ -135,7 +145,28 @@ public virtual string GetVirtualPath(VirtualPathContext context)
...
+
+                int index = url.IndexOfAny(new char[] { '?', '#' });
+                if (index >= 0)
+                {
+                    lowercaseUrl = url.Substring(0, index);
+                    queryString = url.Substring(index);
+                }
+
+                lowercaseUrl = lowercaseUrl.ToLowerInvariant();
+                url = lowercaseUrl + queryString;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouteCollection.cs]: would be good to see a comment here to the effect of:
>  `queryString` will contain the delimiter ? or # as the first character, so it's safe to append.

File: src/Microsoft.AspNet.Routing/RouteOptions.cs [hunk 4]
```diff
@@ -9,6 +9,11 @@ namespace Microsoft.AspNet.Routing
 {
     public class RouteOptions
     {
+        /// <summary>
+        /// Property indicating if generation of lower case URLs is enabled.
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouteOptions.cs]: Clean this up.

> Gets or sets a value indicating whether all generated URLs are lower-case.

File: test/Microsoft.AspNet.Routing.Tests/Constraints/ConstraintsTestHelper.cs [hunk 5]
```diff
@@ -28,6 +29,15 @@ public static bool TestConstraint(IRouteConstraint constraint, object value, Act
             var routeDirection = RouteDirection.IncomingRequest;
             return constraint.Match(context.Object, route, parameterName, values, routeDirection);
         }
+
+        private static IOptions<RouteOptions> GetIOptionsWithRouteOptions(bool lowerCaseUrls = true)
```
Comment (rynowak) [inline test/Microsoft.AspNet.Routing.Tests/Constraints/ConstraintsTestHelper.cs]: `GetIOptionsWithRouteOptions` - We don't generally put the `I` in a name when referring to something in a variable. Just simplify it to `GetOptions`.

File: test/Microsoft.AspNet.Routing.Tests/Constraints/ConstraintsTestHelper.cs [hunk 6]
```diff
@@ -28,6 +29,15 @@ public static bool TestConstraint(IRouteConstraint constraint, object value, Act
...
+        {
+            RouteOptions routeOptions = new RouteOptions();
```
Comment (rynowak) [inline test/Microsoft.AspNet.Routing.Tests/Constraints/ConstraintsTestHelper.cs]: so much var

File: test/Microsoft.AspNet.Routing.Tests/Constraints/ConstraintsTestHelper.cs [hunk 7]
```diff
@@ -28,6 +29,15 @@ public static bool TestConstraint(IRouteConstraint constraint, object value, Act
...
+            routeOptions.LowercaseUrls = lowerCaseUrls;
+            var iOptions = new Mock<IOptions<RouteOptions>>();
```
Comment (rynowak) [inline test/Microsoft.AspNet.Routing.Tests/Constraints/ConstraintsTestHelper.cs]: just call it `optionsAccessor`

File: test/Microsoft.AspNet.Routing.Tests/RouteCollectionTest.cs [hunk 8]
```diff
@@ -16,6 +18,32 @@ namespace Microsoft.AspNet.Routing
...
     public class RouteCollectionTest
     {
+        [Theory]
+        [InlineData(@"Home/Index/23", "home/index/23", true)]
+        [InlineData(@"Home/Index/23", "Home/Index/23", false)]
+        [InlineData(@"Home/Index/23?Param1=ABC&Param2=Xyz", "Home/Index/23?Param1=ABC&Param2=Xyz", false)]
+        [InlineData(@"Home/Index/23?Param1=ABC&Param2=Xyz", "home/index/23?Param1=ABC&Param2=Xyz", true)]
+        [InlineData(@"Home/Index/23#Param1=ABC&Param2=Xyz", "Home/Index/23#Param1=ABC&Param2=Xyz", false)]
+        [InlineData(@"Home/Index/23#Param1=ABC&Param2=Xyz", "home/index/23#Param1=ABC&Param2=Xyz", true)]
+        public void GetVirtualPathTests(string returnUrl, string lowercaseUrl, bool lowercaseUrls)
```
Comment (rynowak) [inline test/Microsoft.AspNet.Routing.Tests/RouteCollectionTest.cs]: needs a better name.

File: test/Microsoft.AspNet.Routing.Tests/RouteCollectionTest.cs [hunk 9]
```diff
@@ -228,7 +259,78 @@ public void NamedRouteTests_GetNamedRoute_AmbiguousRoutesInCollection_ThrowsForA
...
             var ex = Assert.Throws<InvalidOperationException>(() => routeCollection.GetVirtualPath(virtualPathContext));
-            Assert.Equal("The supplied route name 'ambiguousRoute' is ambiguous and matched more than one route.", ex.Message);
+            Assert.Equal(
+                "The supplied route name 'ambiguousRoute' is ambiguous and matched more than one route.", 
+                ex.Message);
+        }
+
+        // "Integration" tests for RouteCollection
+
+        public static IEnumerable<object[]> DataTokens
```
Comment (rynowak) [inline test/Microsoft.AspNet.Routing.Tests/RouteCollectionTest.cs]: what does this mean? `DataTokens` is a defined concept in routing, and I don't think this is it.

**Comment by rynowak** [comment]
:watch:  please rebase on top of dev. The way I did options access in RC was to just grab it from services on execution. If you really like your way better let's talk.

**Comment by kulmugdha** [comment]
@rynowak Inccorporated your comments.

File: src/Microsoft.AspNet.Routing/RouteBuilder.cs [hunk 10]
```diff
@@ -3,6 +3,8 @@
 
 using System;
 using System.Collections.Generic;
+using Microsoft.Framework.DependencyInjection;
+using Microsoft.Framework.OptionsModel;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouteBuilder.cs]: revert

File: src/Microsoft.AspNet.Routing/RouteCollection.cs [hunk 11]
```diff
@@ -197,7 +198,31 @@ public virtual string GetVirtualPath(VirtualPathContext context)
...
-        private void EnsureLogger(HttpContext context)
+        private string NormalizeVirtualPath(String url)
+        {
+            if (string.IsNullOrEmpty(url)) return url;
+            if (_options.LowercaseUrls)
+            {
+                var lowercaseUrl = url;
+                var queryString = "";
+
+                int index = url.IndexOfAny(new char[] { '?', '#' });
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouteCollection.cs]: var

File: test/Microsoft.AspNet.Routing.Tests/Constraints/ConstraintsTestHelper.cs [hunk 12]
```diff
@@ -6,6 +6,7 @@
 using System;
 using System.Collections.Generic;
 using Microsoft.AspNet.Http;
+using Microsoft.Framework.OptionsModel;
```
Comment (rynowak) [inline test/Microsoft.AspNet.Routing.Tests/Constraints/ConstraintsTestHelper.cs]: revert

File: test/Microsoft.AspNet.Routing.Tests/RouteCollectionTest.cs [hunk 13]
```diff
@@ -17,6 +18,35 @@ namespace Microsoft.AspNet.Routing
...
     public class RouteCollectionTest
     {
+        [Theory]
+        [InlineData(@"Home/Index/23", "home/index/23", true)]
+        [InlineData(@"Home/Index/23", "Home/Index/23", false)]
+        [InlineData(@"Home/Index/23?Param1=ABC&Param2=Xyz", "Home/Index/23?Param1=ABC&Param2=Xyz", false)]
+        [InlineData(@"Home/Index/23?Param1=ABC&Param2=Xyz", "home/index/23?Param1=ABC&Param2=Xyz", true)]
+        [InlineData(@"Home/Index/23#Param1=ABC&Param2=Xyz", "Home/Index/23#Param1=ABC&Param2=Xyz", false)]
+        [InlineData(@"Home/Index/23#Param1=ABC&Param2=Xyz", "home/index/23#Param1=ABC&Param2=Xyz", true)]
+        public void GetVirtualPath_CanLowerCaseUrls_BasedOnOptions(
```
Comment (rynowak) [inline test/Microsoft.AspNet.Routing.Tests/RouteCollectionTest.cs]: Please include a case with turkish i. We're depending on invariant culture in this test so let's verify that fact.

File: test/Microsoft.AspNet.Routing.Tests/RouteCollectionTest.cs [hunk 14]
```diff
@@ -697,6 +830,14 @@ private static INamedRouter CreateNamedRoute(string name, bool accept = false, s
 
             return target;
         }
+
+        private static RouteOptions GetRouteOptions(bool lowerCaseUrls = false)
+        {
+            var routeOptions = new RouteOptions();
+            routeOptions.LowercaseUrls = lowerCaseUrls;
+            routeOptions.UseBestEffortLinkGeneration = true;
```
Comment (rynowak) [inline test/Microsoft.AspNet.Routing.Tests/RouteCollectionTest.cs]: might want to make this a parameter as well, as `true` is not the default here.

**Comment by rynowak** [comment]
:watch: this is looking very close to done. Would like to see one test added for unicode goodness and some cleanup.

File: src/Microsoft.AspNet.Routing/RouteCollection.cs [hunk 15]
```diff
@@ -197,7 +198,31 @@ public virtual string GetVirtualPath(VirtualPathContext context)
...
+                if (index >= 0)
```
Comment (Eilon) [inline src/Microsoft.AspNet.Routing/RouteCollection.cs]: Personally I'd find it more readable by doing all the work in one `if()`, basically like this:

```c#
int indexOfSeparator = url.IndexOfAny(...);
if (indexOfSeparator == -1)
{
    // no separator, just lowercase the whole thing
    ...
}
else
{
    // there's a separator, so do some fancy splitting and assembling
    ...
}
```

File: test/Microsoft.AspNet.Routing.Tests/RouteCollectionTest.cs (hunk 14 above)
Comment (kulmugdha) [inline test/Microsoft.AspNet.Routing.Tests/RouteCollectionTest.cs]: Correct, my focus here is lower case urls so I decided to hard code it, but wont hurt to take as parameter so others can use the method if they want to in future.

**Comment by kulmugdha** [comment]
@rynowak Incorporated the comments

File: src/Microsoft.AspNet.Routing/RouteCollection.cs [hunk 16]
```diff
@@ -197,7 +198,35 @@ public virtual string GetVirtualPath(VirtualPathContext context)
...
+                var indexOfSeparator = url.IndexOfAny(new char[] { '?', '#' });
+
+                // No query string, lowercase the url
+                if (indexOfSeparator == -1)
+                {
+                    return url.ToLowerInvariant();
+                }
+                else
+                {
+                    var lowercaseUrl = url.Substring(0, indexOfSeparator);
```
Comment (Eilon) [inline src/Microsoft.AspNet.Routing/RouteCollection.cs]: Can already do the `.ToLowerInvariant()` here:

    var lowercaseUrl = url.Substring(0, indexOfSeparator).ToLowerInvariant();

File: src/Microsoft.AspNet.Routing/RouteCollection.cs [hunk 17]
```diff
@@ -197,7 +198,35 @@ public virtual string GetVirtualPath(VirtualPathContext context)
...
+
+                    lowercaseUrl = lowercaseUrl.ToLowerInvariant();
+
+                    // queryString will contain the delimiter ? or # as the first character, so it's safe to append.
+                    return lowercaseUrl + queryString;
+                }
+            }
+
+            return url;
+    }
```
Comment (Eilon) [inline src/Microsoft.AspNet.Routing/RouteCollection.cs]: Some formatting here got messed up.

Comment (kulmugdha) [inline src/Microsoft.AspNet.Routing/RouteCollection.cs]: fixed both these things.

Comment (Eilon) [inline src/Microsoft.AspNet.Routing/RouteCollection.cs]: (I don't see the formatting fix in the PR, so I assume it's in your local branch.)

**Comment by Eilon** [comment]
Code looks good to me; @rynowak any final thoughts?

File: src/Microsoft.AspNet.Routing/RouteCollection.cs (hunk 17 above)
Comment (kulmugdha) [inline src/Microsoft.AspNet.Routing/RouteCollection.cs]: If you refresh the same commit, you should see both the changes.

**Comment by rynowak** [comment]
:shipit:

**Comment by kulmugdha** [comment]
Thanks @rynowak and @Eilon ! I'll wait to check this in till I fix MVC test failures caused by this.

**Pull request closed without merging**